aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -739,103 +741,30 @@ SparseTensor _coalesce_sparse_cpu(const SparseTensor& self) {
+SparseTensor sparse_mask(const Tensor& t, const SparseTensor& mask) {
   TORCH_CHECK(
       mask.sizes().equals(t.sizes()),
       "sparse_mask: operands have incompatible sizes; self has size ",
       t.sizes(),
       " but mask has size ",
       mask.sizes());
-  AT_ASSERT(!t.is_cuda()); // we were supposed to have dispatched on this
-  TORCH_CHECK(
-      !r.is_cuda(), "sparse_mask: expected 'out' to be CPU, but got CUDA");
-  TORCH_CHECK(
-      !mask.is_cuda(), "sparse_mask: expected 'mask' to be CPU, but got CUDA");
-  at::resize_as_sparse_(r, mask);
-  if (mask._nnz() == 0) {
-    return r.zero_();
-  }
-  int64_t dim = t.dim();
-  int64_t sparse_dim = mask.sparse_dim();
-  Tensor mask_indices = mask._indices();
-  Tensor mask_values = mask._values();
-  Tensor r_values = at::empty(mask_values.sizes(), r._values().options());
-  alias_into_sparse(r, mask_indices.clone(), r_values);
-  r._coalesced_(mask.is_coalesced());
-  int64_t r_nnz = mask._nnz();
-  get_sparse_impl(r)->set_nnz_and_narrow(r_nnz);
-
-  if (t.numel() ==
-      0) { // if t is an empty tensor, there is no need to mask its elements
-    return r;
-  }
-
-  if (dim > sparse_dim) {
-    // Get a flattened sparse indices, similar to NOTE [ Flatten Sparse Indices
-    // ]. Keeping this implementation because it is faster than
-    // flatten_indices()
-    Tensor indices = at::zeros({mask._nnz()}, mask_indices.options());
-    for (const auto d : c10::irange(mask.sparse_dim())) {
-      indices.mul_(mask.size(d));
-      indices.add_(mask_indices.select(0, d));
-    }
 
-    std::vector<int64_t> view_size(1 + mask.dense_dim());
-    view_size[0] = -1;
-    for (const auto d : c10::irange(mask.dense_dim())) {
-      view_size[d + 1] = mask.size(mask.sparse_dim() + d);
-    }
+  if (!mask.is_coalesced()) {
+    TORCH_WARN_ONCE(
+        "sparse_mask(): uncoalesced mask may result in duplicate values in the result. "
+        "It is recommended to pass mask.coalesce() if such behavior is not expected.");
+  }
 
-    Tensor t_view = t.view(view_size);
-    // TODO: Re-audit this; it used to be an indexSelect directly into r_values
-    at::index_select_out(r_values, t_view, 0, indices);
-  } else {
-    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(at::ScalarType::Half, r_values.scalar_type(), "sparse_mask", [&] {
-      sparse_mask_out_cpu_kernel<scalar_t>(
-          r_values, t, r_nnz, sparse_dim, mask_indices);
-    });
+  if (!mask.numel()) {
+    return mask.clone();
   }
-  return r;
-}
 
-SparseTensor sparse_mask_cpu(const Tensor& t, const SparseTensor& mask) {
-  SparseTensor r = at::empty({0}, t.options().layout(kSparse));
-  sparse_mask_out_cpu(r, t, mask);
-  return r;
+  const auto mask_values = mask._values();
+  auto mask_template = at::sparse_coo_tensor(
+      mask._indices(),
+      at::ones({1}, mask_values.options()).expand_as(mask_values),
+      mask.sizes())._coalesced_(mask.is_coalesced());
+  return t.mul(mask_template);
 }
```

Comment:
While we could update the tests for `sparse_mask` with uncoalesced masks, I think it could be redundant given that `mul`  is comprehensively tested.